Fix exit code

tuladhar opened the pull request:
Exit with proper exit code (1) and an error message if elasticsearch executable binary does not exists or has insufficient permissions to execute.

tuladhar committed:
Fix exit code

Exit with proper exit code (1) and an error message if elasticsearch executable binary does not exists or has insufficient permissions to execute.

jasontedor commented on distribution/deb/src/main/packaging/init.d/elasticsearch:
```diff
@@ -109,8 +109,10 @@ export ES_GC_LOG_FILE
-# Check DAEMON exists
-test -x $DAEMON || exit 0
+if [ ! -x "$DAEMON" ] {
+	echo "Could not find elasticsearch executable binary."
```
The test is whether or not the elasticsearch *script* (not a *binary*; standard Java applications do not have executable binaries anyway) exists and is executable. Can you change this to say something along the lines of "the elasticsearch script does not exist or is not executable"?

jasontedor commented:
@tuladhar Thanks for picking this up. The change is good, I left a nitty comment though. I would also like that there is a test for this added to the bats tests. Let me know if you need guidance through this process but you can start by looking in [`qa/vagrant/src/test/resources/packaging/scripts`](https://github.com/elastic/elasticsearch/tree/93f1a5616881c47f8004dc552b8520beee977d14/qa/vagrant/src/test/resources/packaging/scripts).

tuladhar commented:
@jasontedor Requested changes in PR #17219 

Sorry, I haven't worked on bats tests before :(

Here's my try:
```
@test "[INIT.D] ensure elasticsearch startup script exists" {
    assert_file_exist "/usr/share/elasticsearch/bin/elasticsearch"
}
```
How about adding above test in 70_sysv_initd.bats?

Thanks

jasontedor commented:
> Requested changes in PR #17219

No need to open another PR, you can just push new commits to this one.

> Sorry, I haven't worked on bats tests before :(

Neither had I until late last week. There's always a first time, and I'd be happy to work through this one with you. :smile: 

> How about adding above test in 70_sysv_initd.bats?

That's definitely the right place, but your proposal isn't testing the change here. The change here is to exit with an error status if `/usr/share/elasticsearch/bin/elasticsearch` either does not exist or is not executable. So a test of this should set these two situations up and ensure that the init scripts then exit with the expected status code.

tuladhar committed:
Fixed IF syntax and failure message

tuladhar committed:
Add elasticsearch startup script test

The test ensures that elasticsearch startup script exists and is executable.

tuladhar commented:
Thanks @jasontedor for the pointer. I have pushed changes directly to this branch :+1: 

> Neither had I until late last week. There's always a first time, and I'd be happy to work through this one with you. :smile:

Glad to know that. I've checked out [bats github page](https://github.com/sstephenson/bats) and it's fairly straightforward to use. I've pushed changes to `70_sysv_initd.bats` adding elasticsearch startup existence/executable test using `test` command. Please, review.

dakrone commented:
@tuladhar I think to test this a good way to do it would be to `chmod -x` the init file, try starting ES up, verify it doesn't start and exits with a code of 1, then `chmod +x` the file back. What do you think @jasontedor?

jasontedor commented:
@dakrone Yes, that's exactly what I had in mind. We already do something similar in the plugins bats tests.

tuladhar committed:
Improve elasticsearch startup script test

Added improvement as per conversation in https://github.com/elastic/elasticsearch/pull/17082#issuecomment-206459613

tuladhar commented:
Sounds good :+1: 

I looked into few plugin bats tests as @jasontedor mentioned:
* https://github.com/elastic/elasticsearch/blob/0f00c14afc8428a2a72c0b766d2171029dc8f6e1/qa/vagrant/src/test/resources/packaging/scripts/20_tar_package.bats#L76

* https://github.com/elastic/elasticsearch/blob/0f00c14afc8428a2a72c0b766d2171029dc8f6e1/qa/vagrant/src/test/resources/packaging/scripts/20_tar_package.bats#L76

Please review the changes 79a2b03

dakrone commented on qa/vagrant/src/test/resources/packaging/scripts/70_sysv_initd.bats:
```diff
@@ -57,6 +57,18 @@ setup() {
 
+@test "[INIT.D] elasticsearch fails if startup script is not executable" {
+    local INIT="/etc/init.d/elasticsearch"
+    local DAEMON="$ES_HOME/bin/elasticsearch"
```
This needs to be `$ESHOME` (no underscore)

dakrone commented:
Hi @tuladhar I ran these tests today, their is a typo but otherwise they look good, if you want to correct the typo I'll merge this, thanks!

tuladhar committed:
Fixed ES_HOME typo

tuladhar commented:
Thanks @dakrone 

Fixed the typo.

dakrone commented:
Pushed this as a squashed commit: 9da88a99daca0cc490a758956ec92a5fbe69abcf thanks @tuladhar!
